src/callbacks_step/glm_speed.jl

```diff
@@ -0,0 +1,70 @@
+
+"""
+    GlmSpeedCallback(; glm_scale=0.5, cfl=1.0)
+
+Update the divergence cleaning wave speed `c_h` according to the time step
+computed in [`StepsizeCallback`](@ref) for the ideal GLM-MHD equations.
+"""
+struct GlmSpeedCallback{RealT<:Real}
+  glm_scale::RealT
+  cfl::RealT
+end
+
+
+function Base.show(io::IO, cb::DiscreteCallback{Condition,Affect!}) where {Condition, Affect!<:GlmSpeedCallback}
+  glm_speed_callback = cb.affect!
+  @unpack glm_scale, cfl = glm_speed_callback
+  print(io, "GlmSpeedCallback(glm_scale=", glm_scale, ", cfl=", cfl, ")")
+end
+
+
+function Base.show(io::IO, ::MIME"text/plain", cb::DiscreteCallback{Condition,Affect!}) where {Condition, Affect!<:GlmSpeedCallback}
+  if get(io, :compact, false)
+    show(io, cb)
+  else
+    glm_speed_callback = cb.affect!
+
+    setup = [
+             "GLM wave speed scaling" => glm_speed_callback.glm_scale,
+             "Expected CFL number" => glm_speed_callback.cfl,
+            ]
+    summary_box(io, "GlmSpeedCallback", setup)
+  end
+end
+
+
+function GlmSpeedCallback(; glm_scale=0.5, cfl=1.0)
+  # when is the callback activated
+  condition = (u, t, integrator) -> true
+
+  glm_speed_callback = GlmSpeedCallback(glm_scale, cfl)
+
+  DiscreteCallback(condition, glm_speed_callback,
+                   save_positions=(false,false),
+                   initialize=initialize!)
+end
+
+
+function initialize!(cb::DiscreteCallback{Condition,Affect!}, u, t, integrator) where {Condition, Affect!<:GlmSpeedCallback}
+  cb.affect!(integrator)
+end
+
+
+# This method is called as callback after the StepsizeCallback during the time integration.
+@inline function (glm_speed_callback::GlmSpeedCallback)(integrator)
+
+  dt = get_proposed_dt(integrator)
+  semi = integrator.p
+  _, equations, solver, cache = mesh_equations_solver_cache(semi)
+  @unpack glm_scale, cfl = glm_speed_callback
+
+  # compute time step for GLM linear advection equation with c_h=1 (redone due to the possible AMR)
+  c_h_deltat = calc_dt_for_cleaning_speed(cfl, equations, solver, cache)
+
+  # c_h is proportional to its own time step divided by the complete MHD time step
+  equations.c_h = glm_scale * c_h_deltat / dt
```

Comment:
I wonder what happens if the callback is not created (or  not created properly). What will the value of `c_h` value be? Is it a usable default such that the user won't even notice? My thinking here is as follows: If there is a useful default value, then it should be initialized accordingly and we just have to document it properly that users are aware of this. OTOH, if there is no sensible default, I'd rather set the speed to something that makes the solver fail fast & hard, e.g., `NaN` or a negative value.

Optimally, we would have some kind of check that stops the run when MHD is used without the callback, but there is no place (that I know of) where this kind of functionality could be implemented.

Reply:
The "safety" value that is drops to without this callback should (optimally) be the global value of the fast magnetoacoustic wave speed over the entire domain. This is probably complicated and fairly "hacky" to put in. We can have it default to 0 without the callback, then GLM is basically turned off. The code would (usually) run but the results might look bad because of pollution due to unphysical growth in the div(B) and this, e.g., influences the magnetic energy.

As you said, it would be good if we could throw an error if one forgets this callback.

Reply:
I don't see a "clean" solution to throw an error if the callback is forgotten, but maybe @ranocha has an idea?

Other than that, I think *you* are best qualified to decide about a sensible default value. If there are some cases where it makes sense to run without the callback, but usually it is an error, maybe something like this: Modify the MHD equation constructor from
https://github.com/trixi-framework/Trixi.jl/blob/4722d531419d846011fce9aa4f93e14192df52ec/src/equations/ideal_glm_mhd_2d.jl#L12-L14
to
```julia
function IdealGlmMhdEquations2D(gamma; initial_c_h=NaN)
  IdealGlmMhdEquations2D(promote(gamma, initial_c_h)...)
end
```
This way, `c_h` must be set to a useful value by the callback, otherwise the simulation crashes immediately. If a user decides, for whatever reason, to use run an MHD simulation without the `GlmSpeedCallback`, they can always create their equations as `IdealGlmMhdEquations2D(1.4, 0.0)` to manually set it to zero. Would that make sense?

Reply:
I don't think there is any "clean" way to "force" users to use the `GlmSpeedCallback` for the MHD equations - in some sense, that would defeat the purpose of Trixi as a *library* that can be applied to setup simulations as users like. You could hack something, e.g. looking for a `GlmSpeedCallback` when initializing the `StepsizeCallback` - and vice versa, since that's an implicit assumption of the `GlmSpeedCallback` (and users can choose to run with fixed time steps or error-base step size control).

Concerning the constructor: You should better use
```julia
function IdealGlmMhdEquations2D(gamma; initial_c_h=convert(typeof(gamma), NaN))
  # Use `promote` to ensure that `gamma` and `initial_c_h` have the same type
  IdealGlmMhdEquations2D(promote(gamma, initial_c_h)...)
end
```
since, `gamma` will be promoted to `Float64` by default in the version above, which might not be desired.

Reply:
OK, I agree. We should not (yet 😉 ) start hacking setup-specific stuff into the callbacks, at least if it's "only" for convenience.

And thanks for spotting this type bug in my code, I really need to be more careful about stuff like this. I also tend to forget that in Julia you can use argument information from arguments to the left...